cmd/soroban-rpc/internal/db/db.go

```diff
@@ -128,33 +128,29 @@ func getMetaValue(ctx context.Context, q db.SessionInterface, key string) (strin
 	case 1:
 		// expected length on an initialized DB
 	default:
-		return "", fmt.Errorf("multiple entries (%d) for key %q in table %q", len(results), latestLedgerSequenceMetaKey, metaTableName)
+		return "", fmt.Errorf("multiple entries (%d) for key %q in table %q", len(results), key, metaTableName)
 	}
 	return results[0], nil
 }
 
-func getLatestLedgerSequence(ctx context.Context, q db.SessionInterface, cache *dbCache) (uint32, error) {
-	latestLedgerStr, err := getMetaValue(ctx, q, latestLedgerSequenceMetaKey)
+func getLatestLedgerSequence(ctx context.Context, ledgerReader LedgerReader, cache *dbCache) (uint32, error) {
+	ledgerRange, err := ledgerReader.GetLedgerRange(ctx)
```

Comment:
this call will end up querying the DB even if the cache is populated. I think we should first check the cache and return that value if it exists. otherwise, we can query the db and populate the cache

Reply:
@tamirms This function only gets called during filling of entries via the checkpoint - https://github.com/stellar/soroban-rpc/blob/main/cmd/soroban-rpc/internal/ingest/service.go#L175. At all other instances, we have a check on the cache to see if its populated and return that value instead:
- https://github.com/stellar/soroban-rpc/blob/main/cmd/soroban-rpc/internal/db/ledgerentry.go#L148
- https://github.com/stellar/soroban-rpc/pull/256/files#diff-c57d2d5677168ee9d46b7a74fb7a34a5ba6e11448ada5854e6c85d4246140fa2R345

I tried adding a check directly where you pointed, but it leads to race condition issues during concurrent reads and writes. @2opremio was also mentioning encountering this when he added this change first.

Reply:
> At all other instances, we have a check on the cache to see if its populated and return that value instead

The fact that this check is repeated more than once is a sign that the check should be encapsulated in `getLatestLedgerSequence()` so that all the callers of `getLatestLedgerSequence()` don't have to check the cache.

> I tried adding a check directly where you pointed, but it leads to race condition issues during concurrent reads and writes

if there's no way to resolve this issue we can keep the code as is. Could you describe the race condition in more detail?

Reply:
@tamirms It was a mistake on my side. I was not locking the caching during read and that was leading to the race condition. It now works fine and the tests are passing